pubsub/gochannel/pubsub.go

```diff
@@ -344,8 +356,16 @@ func (s *subscriber) sendMessageToSubscriber(msg *message.Message, logFields wat
 	s.sending.Lock()
 	defer s.sending.Unlock()
 
-	ctx, cancelCtx := context.WithCancel(s.ctx)
-	defer cancelCtx()
+	var ctx context.Context
+
+	//This is getting the context from the message, not the subscriber
+	if s.preserveContext {
+		ctx = msg.Context()
+	} else {
+		var cancelCtx context.CancelFunc
+		ctx, cancelCtx = context.WithCancel(s.ctx)
+		defer cancelCtx()
+	}
```

Comment:
I would like to suggest this


```suggestion
	ctx := msg.Context()

	//This is getting the context from the message, not the subscriber
	if !s.preserveContext {
		var cancelCtx context.CancelFunc
		ctx, cancelCtx = context.WithCancel(s.ctx)
		defer cancelCtx()
	}
```

Reply:
Agreed, changed

A quick question though :-

When we are using the subscriber context, even if the subscriber main context is cancelled, that time the messages will still be processed further.
So why are we even using cancelCtx here.

Reply:
To be honest, I was surprised to see an option to change it, for me the context should be always the one of the message

The lib is in charge to pass the context to the message, so if the context is cancelled, the message one is too, no ?

I mean couldn't we simply fix it, and not adding the preserveContext boolean ?

Reply:
I agree, however @m110 has a point here - 
https://github.com/ThreeDotsLabs/watermill/issues/464#issuecomment-2283885824

Nonetheless, according to your statement - 
`The lib is in charge to pass the context to the message, so if the context is cancelled, the message one is too, no ?
`

I don't see that happening in the code though, even if the main context is cancelled, the messages are passed to the subscriber nonetheless

Reply:
Yes, they are, but the context of the message would be canceled I think/hope

Anyway this comment might change everything https://github.com/ThreeDotsLabs/watermill/pull/487#discussion_r1749848993

So, let's wait